draft-ietf-tls-esni.md

```diff
@@ -723,38 +716,56 @@ error code.
+then the client presumes rejection and proceeds as if the message was sent by
+the client-facing server.
+
+The HelloRetryRequest constructed by the backend server is called the
+HelloRetryRequestInner message; and the HelloRetryRequest constructed by the
+client-facing server is called the HelloRetryRequestOuter message.
+
+If the client offered ECH in the first ClientHello, then it MUST offer ECH in
+the second. Likewise, if the client did not offer ECH in the first ClientHello,
+then it MUST NOT not offer ECH in the second.
+
+#### Accepted ECH {#client-hrr-accepted-ech}
 
-1. TLS 1.3 {{!RFC8446}} ciphersuites in the ClientHello.cipher_suites list.
-1. The "key_share" and "supported_groups" extensions {{RFC8446}}. (These
-extensions may be copied from ClientHelloOuter into ClientHelloInner as
-described in {{real-ech}}.)
-1. Versions in the "supported_versions" extension, excluding TLS 1.2 and
-earlier. Note the ClientHelloOuter MAY include these older versions, while the
-ClientHelloInner MUST omit them.
+As described in {{sever-hrr}, the HelloRetryRequestInner handshake message is
+encrypted by the backend server and included as the payload of the
+"encrypted_client_hello" extension of the HelloRetryRequestOuter. The client
+decrypts the payload using the HPKE context it shares with the client-facing
+server. It begins by exporting a key and nonce for AEAD decryption:
 
-Future extensions that might change across first and second ClientHello messages
-in response to a HelloRetryRequest MUST have the same value.
+~~~
+    hrr_inner_key = context.Export("hrr inner key", Nk)
+    hrr_inner_nonce = context.Export("hrr inner nonce", Nn)
+~~~
 
-If the server sends a HelloRetryRequest in response to the ClientHello, the
-client sends a second updated ClientHello per the rules in {{RFC8446}}.
-However, at this point, the client does not know whether the server processed
-ClientHelloOuter or ClientHelloInner, and MUST regenerate both values to be
-acceptable. Note: if ClientHelloOuter and ClientHelloInner use different groups
-for their key shares or differ in some other way, then the HelloRetryRequest
-may actually be invalid for one or the other ClientHello, in which case a fresh
-ClientHello MUST be generated, ignoring the instructions in HelloRetryRequest.
-Otherwise, the usual rules for HelloRetryRequest processing apply.
-
-The client encodes the second ClientHelloInner as in {{encoding-inner}}, using
-the second ClientHelloOuter for any referenced extensions. It then encrypts
-the new EncodedClientHelloInner value as a second message with the previous
-HPKE context:
+Here `Nk` and `Nn` denote, respectively, the key and nonce length of the AEAD
+algorithm associated with `context`. Let `encrypted_hrr_inner` denote the
+extension payload. The client computes
+
+~~~
+    HelloRetryRequestInner = Open(hrr_inner_key,
+                                  hrr_inner_nonce,
+                                  "", // Empty AAD
```

Comment:
Is there a reason not to AAD HRROuter as with CH?

Reply:
No reason that I see, other than a bit more code complexity. This seems like a sensible precaution to me, so I added it.